src/order/complete_well_founded.lean

```diff
@@ -0,0 +1,156 @@
+      have h_Sup : has_Sup.Sup s = has_Sup.Sup s',
+      { apply le_antisymm,
+        { suffices : s ⊆ s', { exact Sup_le_Sup this, },
+          intros x hx,
+          use { x }, rw Sup_singleton, simp [hx], },
+        { have hs' : ∀ (x ∈ s'), x ≤ Sup s,
+          { rintros x ⟨t, h_subset, h_finite, hx⟩, rw hx, exact Sup_le_Sup h_subset, },
+          apply Sup_le, exact hs', }, },
+      have hs' : ∀ a b, a ∈ s' → b ∈ s' → a ⊔ b ∈ s',
+      { rintros a b ⟨t₁, h_subset₁, h_finite₁, ha⟩ ⟨t₂, h_subset₂, h_finite₂, hb⟩,
+        use t₁ ∪ t₂,
+        refine ⟨set.union_subset h_subset₁ h_subset₂, set.finite.union h_finite₁ h_finite₂, _⟩,
+        rw [ha, hb, Sup_union.symm], },
+      have hne' : s'.nonempty,
+      { let t : set α := { hne.some },
+        have ht : t.finite, { exact set.finite_singleton (set.nonempty.some hne), },
+        use Sup t, refine ⟨t, _, ht, rfl⟩,
+        simp only [set.singleton_subset_iff], apply hne.some_spec, },
+      rw h_Sup, obtain ⟨t, ht, h_finite, h_Sup'⟩ := h s' hne' hs',
+      use t, split,
+      { use ht, },
+      { use h_finite, exact h_Sup', }, },
+    { use ⊥, split,
+      { simp only [set.empty_subset, set.bot_eq_empty], },
+      { simp only [set.bot_eq_empty, Sup_empty, Sup_eq_bot, set.finite_empty, exists_prop_of_true],
+        intros a ha, exfalso, apply hne, exact ⟨a, ha⟩, }, },  },
+  { suffices : ∀ (s : set α), (⊥ ∈ s) → (∀ a b, a ∈ s → b ∈ s → a ⊔ b ∈ s) → (has_Sup.Sup s) ∈ s,
+    { intros s hne h_sup,
+      have h_sup' : (∀ a b, a ∈ s.insert ⊥ → b ∈ s.insert ⊥ → a ⊔ b ∈ s.insert ⊥),
+      { intros a b ha hb,
+        have ha' := set.eq_or_mem_of_mem_insert ha,
+        have hb' := set.eq_or_mem_of_mem_insert hb,
+        rcases ha'; rcases hb',
+        { simp only [ha', hb', sup_bot_eq], exact s.mem_insert ⊥, },
+        { simp only [ha', hb', bot_sup_eq], exact hb, },
+        { simp only [ha', hb', sup_bot_eq], exact ha, },
+        { apply set.mem_insert_of_mem ⊥, apply h_sup _ _ ha' hb', }, },
+      specialize this (s.insert ⊥) (s.mem_insert ⊥) h_sup',
+      have h₁ : has_Sup.Sup (s.insert ⊥) = has_Sup.Sup s,
+      { change has_Sup.Sup (insert ⊥ s) = has_Sup.Sup s, simp, }, rw h₁ at this,
+      rcases set.eq_or_mem_of_mem_insert this,
+      { have h₂ := Sup_eq_bot.mp h_1, change has_Sup.Sup s ∈ s, rw h_1,
+        rw ← h₂ hne.some hne.some_spec, apply hne.some_spec, },
+      { assumption, }, },
+    intros s h_bot h_sup, obtain ⟨t, h_subset, h_finite, h_Sup⟩ := h s, change Sup s ∈ s, rw h_Sup,
+    revert h_subset, apply h_finite.induction_on,
+    { change _ → has_Sup.Sup ∅ ∈ s, simp [h_bot], },
+    { intros x t' h₁ h₂ h₃ h₄, rw set.insert_subset at h₄, change has_Sup.Sup _ ∈ s,
+      simp only [Sup_insert], exact h_sup _ _ h₄.1 (h₃ h₄.2), }, },
+end
+
+/-- Surely this exists somewhere (at least implicitly) already. -/
+lemma nat.finite_subset_is_bounded {s : set ℕ} (h : s.finite) : ∃ n, ∀ x ∈ s, x ≤ n :=
```

Comment:
I bet you can do this for any `semilattice_sup` with `finset.sup`.

Reply:
Great point, you are of course correct:
```lean
lemma finite_subset_is_bounded {β : Type*} [semilattice_sup_bot β] {s : set β} (h : s.finite) :
  ∃ y, ∀ x ∈ s, x ≤ y :=
begin
  use h.to_finset.sup id,
  intros x hx, exact finset.le_sup (set.finite.mem_to_finset.mpr hx),
end
```

Thanks for such helpful remarks here and elsewhere. I finally have time again this week and plan to rework this PR substantially tomorrow, using your suggestions.